File: packages/sdk-python/voltx_sdk/resources/webhook_endpoints.py
from __future__ import annotations

from typing import TYPE_CHECKING, List

from ..generated.schema import CreateWebhookEndpointDto, UpdateWebhookEndpointDto
from ..types import (
    CreateWebhookEndpointResult,
    RotateWebhookEndpointSecretResult,
    WebhookDelivery,
    WebhookEndpoint,
)

if TYPE_CHECKING:
    from ..client import VoltxClient


class WebhookEndpointsResource:
    def __init__(self, client: "VoltxClient") -> None:
        self._client = client

    def list(self, organization_id: str) -> List[WebhookEndpoint]:
        return self._client.get(f"/organizations/{organization_id}/webhook-endpoints")

    def create(self, organization_id: str, input: CreateWebhookEndpointDto) -> CreateWebhookEndpointResult:
        return self._client.post(f"/organizations/{organization_id}/webhook-endpoints", input)

    def get(self, organization_id: str, id: str) -> WebhookEndpoint:
        return self._client.get(f"/organizations/{organization_id}/webhook-endpoints/{id}")

    def update(self, organization_id: str, id: str, input: UpdateWebhookEndpointDto) -> WebhookEndpoint:
        return self._client.patch(f"/organizations/{organization_id}/webhook-endpoints/{id}", input)

    def rotate_secret(self, organization_id: str, id: str) -> RotateWebhookEndpointSecretResult:
        return self._client.post(f"/organizations/{organization_id}/webhook-endpoints/{id}/rotate-secret")

    def suspend(self, organization_id: str, id: str) -> WebhookEndpoint:
        return self._client.post(f"/organizations/{organization_id}/webhook-endpoints/{id}/suspend")

    def reactivate(self, organization_id: str, id: str) -> WebhookEndpoint:
        return self._client.post(f"/organizations/{organization_id}/webhook-endpoints/{id}/reactivate")

    def delete(self, organization_id: str, id: str) -> None:
        self._client.delete(f"/organizations/{organization_id}/webhook-endpoints/{id}")

    def list_deliveries(self, organization_id: str, id: str) -> List[WebhookDelivery]:
        return self._client.get(f"/organizations/{organization_id}/webhook-endpoints/{id}/deliveries")

    def replay_delivery(self, organization_id: str, id: str, delivery_id: str) -> WebhookDelivery:
        return self._client.post(
            f"/organizations/{organization_id}/webhook-endpoints/{id}/deliveries/{delivery_id}/replay"
        )

File: packages/sdk-typescript/src/resources/webhook-endpoints.ts
import type { VoltxClient } from "../client.js";
import type { components } from "../generated/schema.js";
import type {
  CreateWebhookEndpointResult,
  RotateWebhookEndpointSecretResult,
  WebhookDelivery,
  WebhookEndpoint,
} from "../types.js";

export type CreateWebhookEndpointInput = components["schemas"]["CreateWebhookEndpointDto"];
export type UpdateWebhookEndpointInput = components["schemas"]["UpdateWebhookEndpointDto"];

export class WebhookEndpointsResource {
  constructor(private readonly client: VoltxClient) {}

  list(organizationId: string): Promise<WebhookEndpoint[]> {
    return this.client.get<WebhookEndpoint[]>(`/organizations/${organizationId}/webhook-endpoints`);
  }

  create(organizationId: string, input: CreateWebhookEndpointInput): Promise<CreateWebhookEndpointResult> {
    return this.client.post<CreateWebhookEndpointResult>(
      `/organizations/${organizationId}/webhook-endpoints`,
      input,
    );
  }

  get(organizationId: string, id: string): Promise<WebhookEndpoint> {
    return this.client.get<WebhookEndpoint>(`/organizations/${organizationId}/webhook-endpoints/${id}`);
  }

  update(organizationId: string, id: string, input: UpdateWebhookEndpointInput): Promise<WebhookEndpoint> {
    return this.client.patch<WebhookEndpoint>(
      `/organizations/${organizationId}/webhook-endpoints/${id}`,
      input,
    );
  }

  rotateSecret(organizationId: string, id: string): Promise<RotateWebhookEndpointSecretResult> {
    return this.client.post<RotateWebhookEndpointSecretResult>(
      `/organizations/${organizationId}/webhook-endpoints/${id}/rotate-secret`,
    );
  }

  suspend(organizationId: string, id: string): Promise<WebhookEndpoint> {
    return this.client.post<WebhookEndpoint>(
      `/organizations/${organizationId}/webhook-endpoints/${id}/suspend`,
    );
  }

  reactivate(organizationId: string, id: string): Promise<WebhookEndpoint> {
    return this.client.post<WebhookEndpoint>(
      `/organizations/${organizationId}/webhook-endpoints/${id}/reactivate`,
    );
  }

  delete(organizationId: string, id: string): Promise<void> {
    return this.client.delete<void>(`/organizations/${organizationId}/webhook-endpoints/${id}`);
  }

  listDeliveries(organizationId: string, id: string): Promise<WebhookDelivery[]> {
    return this.client.get<WebhookDelivery[]>(
      `/organizations/${organizationId}/webhook-endpoints/${id}/deliveries`,
    );
  }

  replayDelivery(organizationId: string, id: string, deliveryId: string): Promise<WebhookDelivery> {
    return this.client.post<WebhookDelivery>(
      `/organizations/${organizationId}/webhook-endpoints/${id}/deliveries/${deliveryId}/replay`,
    );
  }
}
